### PR TITLE
Better choices of what pods to kill

### PR DESCRIPTION
**Pull request by janetkuo**
Addresses #22065
- Prefer to kill pods that have been ready for less time (use transition time, similar to how availability is determined)
- Prefer to kill pods with containers with higher restart counts
- If all else is equal, kill newer pods (by creation timestamp) -- which is desirable in the case of overshooting, anyway

Also make the pod-killing preference more predictable. 

@bgrant0607 @kargakis @madhusudancs @mqliang @nikhiljindal @yujuhong

### CODE REVIEW

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -409,22 +410,59 @@ func (s ActivePods) Len() int      { return len(s) }
...
+func podReadyTime(pod *api.Pod) unversioned.Time {
+	if api.IsPodReady(pod) {
+		for _, c := range pod.Status.Conditions {
+			// we only care about pod ready conditions
+			if c.Type == api.PodReady && c.Status == api.ConditionTrue {
+				return c.LastTransitionTime
+			}
+		}
+	}
+	return unversioned.Time{}
```
Comment (bgrant0607) [inline pkg/controller/controller_utils.go]: Is empty time 0 or infinity with respect to After/Before comparisons?

**Comment by bgrant0607** [comment]
LGTM, thanks!

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (janetkuo) [inline pkg/controller/controller_utils.go]: Empty time 0 is always before non-empty time (and vise versa). Should check if the time is empty in step 4.

Comment (janetkuo) [inline pkg/controller/controller_utils.go]: @bgrant0607 Done in the 2nd commit. PTAL

**Comment by bgrant0607** [comment]
LGTM, thanks

**Comment by bgrant0607** [comment]
@k8s-bot test this issue: #22045

**Comment by bgrant0607** [comment]
In the failed e2e tests (kubernetes-jenkins/pr-logs/pull/22078/kubernetes-pull-build-test-e2e-gce/30769), one pod failed to terminate and/or start, and one failed to schedule. Seems totally unrelated to this test.

A quick search didn't find issues filed for the specific tests that failed, but I suspect general node instability, which has been a problem lately.

**Comment by bgrant0607** [comment]
So many flakes, so little time. I'll try to look at them later:
kubernetes-jenkins/pr-logs/pull/22078/kubernetes-pull-build-test-e2e-gce/30769

@k8s-bot test this issue: #IGNORE

**Pull request merged**